src/jsrm/utils/lie_operators.py

```diff
@@ -0,0 +1,297 @@
+    x = s-l_nprev  # Compute the segment length
+    adjoint_xi_n = adjoint_SE3(xi_n)  # Adjoint representation of the input vector
+    
+    Tangent = (x*jnp.eye(6)
+            + 1/(2*jnp.power(theta, 2)) * (
+                4 - 4*jnp.cos(x*theta) - x*theta*jnp.sin(x*theta)
+                ) * adjoint_xi_n
+            + 1/(2*jnp.power(theta, 3)) * (
+                4*x*theta - 5*jnp.sin(x*theta) + x*theta*jnp.cos(x*theta)
+                ) * jnp.linalg.matrix_power(adjoint_xi_n, 2)
+            + 1/(2*jnp.power(theta, 4)) * (
+                2 - 2*jnp.cos(x*theta) - x*theta*jnp.sin(x*theta)
+                ) * jnp.linalg.matrix_power(adjoint_xi_n, 3)
+            + 1/(2*jnp.power(theta, 5)) * (
+                2*x*theta - 3*jnp.sin(x*theta) + x*theta*jnp.cos(x*theta)
+                ) * jnp.linalg.matrix_power(adjoint_xi_n, 4))
+
+    return Tangent
+
+@jit
+def vec_SE2_to_xi_SE3(
+    vec3: Array, 
+    indices: Sequence[int] = (2, 3, 4)
+) -> Array:
+    """
+    Convert a strain vector in se(2) to a strain vector in se(3).
+
+    Args:
+        vec3 (Array): array-like, shape (3,1)
+            A 3-dimensional vector representing the strain in se(2).
+            The first element correspond to the angular component, 
+            and the last elements corresponds to the linear component.
+        indices (Sequence[int], optional): Indices in the 6D se(3) vector 
+            where to insert the se(2) components. Default is (2, 3, 4)
+
+    Returns:
+        Array: shape (6,1)
+            A 6-dimensional vector representing the strain in se(3).
+            The first three elements correspond to the angular component, 
+            and the last three elements correspond to the linear component.
+    """
+    vec3 = jnp.asarray(vec3).flatten()  # Ensure vec3 is a JAX array
+    
+    xi = jnp.zeros((6, ))  # Initialize a 6D vector with zeros
+    xi = xi.at[jnp.array(indices)].set(vec3)  # Set the values at the specified indices
+    return xi.reshape((6, 1))
+
+@jit
+def compute_weighted_sums(
+    M: Array,
+    vec6: Array,
+    idx: int
+)-> Array:
+    """
+    Compute the weighted sums of the matrix product of M and vec6,
+
+    Args:
+        M (Array): array of shape (N, 6, 6)
+           Describes the matrix to be multiplied with vec6
+        vec6 (Array): array-like of shape (N, 6)
+           Describes the vector to be multiplied with M
+        idx (int): index of the last row to be summed over
+
+    Returns:
+        Array: array of shape (N, 6)
+           The result of the weighted sums. For each i, the result is the sum of the products of M[i, j] and vec6[j] for j from 0 to idx.
+    """
+    N = M.shape[0]
+    # Matrix product for each j: (N, 6, 6) @ (N, 6, 1) -> (N, 6)
+    prod = jnp.einsum('nij,nj->ni', M, vec6)
+
+    # Triangular mask for partial sum: (N, N)
+    # mask[i, j] = 1 if j >= i and j <= idx
+    mask = (jnp.arange(N)[:, None] <= jnp.arange(N)[None, :]) & (jnp.arange(N)[None, :] <= idx)
```

Comment:
I don't know if this is the case, but when you use `jnp.arange(N)` and `N` changes (because you maybe call this function with matrices `M` of different dimensionalities), then this will trigger a re-compile which can slow down things a lot.

Please use the `jax.log_compiles` context manger (https://docs.jax.dev/en/latest/_autosummary/jax.log_compiles.html) or the `JAX_LOG_COMPILES=1` (https://docs.jax.dev/en/latest/config_options.html) environment variable to check for recompiles during evaluation.